crates/forge/src/runner.rs

```diff
@@ -707,6 +719,32 @@ impl<'a> ContractRunner<'a> {
             gas_report_traces: result.gas_report_traces.into_iter().map(|t| vec![t]).collect(),
         }
     }
+
+    /// Returns the environment of the test runner
+    ///
+    /// If the backend has forks, return the block number of the fork.
+    /// If the backend does not have forks, return [TestEnvironment::Standard].
+    fn get_environment(&self, env: Option<EnvWithHandlerCfg>) -> TestEnvironment {
+        // If a fork occurs inside of a test, use the block number of the fork as the test
+        // Note: if a user uses `vm.roll` in a test, the test will be marked as a fork.
+        if let Some(environment) = env {
+            let block_number = environment.env.block.number.to::<u64>();
+
+            // If the block number is at genesis we assume it is a non-forked environment and
+            // continue checking for forks in the backend.
+            if block_number > 1 {
+                return TestEnvironment::Fork { block_number }
+            }
+        }
```

Comment:
@mattsse in order to be able to check for forks occuring inside of tests we cannot rely on the executor's backend.

Whilst this current implementation works across different test cases (as in works correctly with `vm.createSelectFork` where it previously did not) the block number check is very rudimentary. Do you see any way to make this more robust?

Another issue is whenever `vm.roll` is used we see (local) forks being created and it gets misclassified. Is there a good way to differentiate between `remote forks` and `local forks`.

Handling all these different edge cases is becoming whack-a-mole and I think time is better spend towards fixing the cache busting issue described here: https://github.com/foundry-rs/foundry/issues/6792#issuecomment-2037800287. When that works as intended the issue will be much less noticeable as users won't run into rate limits as often.

Reply:
Implemented the fix for the cache busting here: https://github.com/foundry-rs/block-explorers/pull/40

Makes a huge difference in performance on repeated runs.

Reply:
Made some modifications, now collecting the context on `Executor::call`

Still relying on checking for non-genesis blocks as a proxy to differentiate between default and fork